### PR TITLE
feature #543: created endpoint to update event

### LINKED ISSUE #543: Create an endpoint to update an existing event
Your task is to update the event entity for a specific index in the schedule. What actually needs to be updated? Let's take a look at the design:

image

As we can see, we need to update the name of the event, its type (if it is a type of discipline, then we need to additionally update all teachers by teacher id). Next, we update the start and end dates, period and conference link.

**Important:** we also need to update the event description, as well as the discipline description (if necessary), because it is the same for all pairs of this event in the schedule.

[Design](https://www.figma.com/file/FHBiKogMc9VzGWojaQxA9c/%F0%9F%94%A52.0?type=design&node-id=242-2&mode=design&t=vWrhHqEHOvNsBSG9-0)

### PR DESCRIPTION
**Pull request by f-ogoog**
- created new endpoint to update event
- added UpdateEventDTO, EventPipe

closes #543

### CODE REVIEW

File: src/v2/api/dtos/UpdateEventDTO.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  @ApiPropertyOptional()
+  @IsOptional()
+  @IsUrl(undefined, validationOptionsMsg('Url must be a URL address'))
+    url?: string;
+
+  @ApiPropertyOptional()
+  @IsOptional()
+  @MaxLength(2000, validationOptionsMsg('Discipline description is too long (max: 2000)'))
+    disciplineDescription?: string;
```
Comment (just-hoshinon) [inline src/v2/api/dtos/UpdateEventDTO.ts:60]: disciplineInfo

File: src/v2/api/dtos/UpdateEventDTO.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  @ApiPropertyOptional()
+  @IsOptional()
+  @MaxLength(2000, validationOptionsMsg('Event description is too long (max: 2000)'))
+    pairDescription?: string;
```
Comment (just-hoshinon) [inline src/v2/api/dtos/UpdateEventDTO.ts:65]: eventInfo

File: src/v2/api/dtos/UpdateEventDTO.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  @IsDate(validationOptionsMsg('Start time must be Date'))
+    startTime?: Date;
+
+  @ApiPropertyOptional()
+  @IsOptional()
+  @Type(() => Date)
+  @IsDate(validationOptionsMsg('Start time must be Date'))
+    endTime?: Date;
+
+  @ApiPropertyOptional()
```
Comment (just-hoshinon) [inline src/v2/api/dtos/UpdateEventDTO.ts:47]: add enum recognition

File: src/v2/api/services/ScheduleService.ts [hunk 4]
```diff
@@ -237,4 +244,175 @@ export class ScheduleService {
...
+        createMany: {
+          data: [],
+        },
+      },
+    };
+
+    update.description = disciplineDescription ? disciplineDescription : undefined;
+
+    if (type) {
+      update.disciplineTypes = {
```
Comment (just-hoshinon) [inline src/v2/api/services/ScheduleService.ts:330]: It's not correct. Changing the type doesn't change the discipline type entity. It disconnects and connects another discipline type

File: src/v2/api/services/ScheduleService.ts [hunk 5]
```diff
@@ -237,4 +244,175 @@ export class ScheduleService {
...
+          data: {
+            name: type,
+          },
+        },
+      };
+    }
+
+    if (teachers) {
+      disciplineTeachers.map((disciplineTeacher) => {
+        if (!teachers.some((t) => disciplineTeacher.teacherId === t)) {
```
Comment (just-hoshinon) [inline src/v2/api/services/ScheduleService.ts:344]: We have util method for such purposes

**Review by just-hoshinon — CHANGES_REQUESTED** [review]

**Review by kropli — CHANGES_REQUESTED** [review]

**Review by kropli — APPROVED** [review]

File: src/v2/api/services/ScheduleService.ts [hunk 6]
```diff
@@ -237,4 +244,228 @@ export class ScheduleService {
...
+        deleteMany: {
+          OR: [],
+        },
+        createMany: {
+          data: [],
+        },
+      },
+    };
+
+    update.description = disciplineInfo ? disciplineInfo : undefined;
```
Comment (Hahhhayam) [inline src/v2/api/services/ScheduleService.ts:329]: use null-union ??

**Review by Hahhhayam — CHANGES_REQUESTED** [review]

File: src/v2/api/services/ScheduleService.ts [hunk 7]
```diff
@@ -322,4 +323,228 @@ export class ScheduleService {
...
+
+    if (period === Period.NO_PERIOD) {
+      data.deleteMany = {
+        number: {
+          gte: 1,
+        },
+      };
+    } else {
+      const countOfWeek = Math.ceil((endTime.getTime() - startTime.getTime()) / WEEK);
+      const countOfPair = period === Period.EVERY_FORTNIGHT ? Math.floor(countOfWeek/2) : countOfWeek;
```
Comment (basarabst) [inline src/v2/api/services/ScheduleService.ts:369]: spaces " /  "

File: src/v2/api/services/ScheduleService.ts [hunk 8]
```diff
@@ -322,4 +323,228 @@ export class ScheduleService {
...
+      const pairDifference = countOfPair - event.eventInfo.length;
+
+      if (pairDifference > 0) {
+        data.createMany = {
+          data: this.scheduleParser.getIndexesForEventInfo(event.eventInfo.length, event.eventInfo.length+pairDifference-1),
```
Comment (basarabst) [inline src/v2/api/services/ScheduleService.ts:374]: spaces

File: src/v2/api/services/ScheduleService.ts [hunk 9]
```diff
@@ -322,4 +323,228 @@ export class ScheduleService {
...
+        };
+      } else if (pairDifference < 0) {
+        data.deleteMany = {
+          number: {
+            gte: countOfPair-1,
```
Comment (basarabst) [inline src/v2/api/services/ScheduleService.ts:379]: spaces

File: src/v2/api/services/ScheduleService.ts [hunk 10]
```diff
@@ -322,4 +323,228 @@ export class ScheduleService {
...
+    const period = body.period ?? event.period;
+    const eventInfoForUpdate = await this.getEventInfoForUpdate(period, startTime, endTime, week, eventInfo, event);
+
+    const newEvent = await this.eventRepository.updateById(eventId, {
+      ...update,
+      eventInfo: eventInfoForUpdate,
+    });
+
+    if (disciplineId) {
+      if (!type || !teachers) throw new ObjectIsRequiredException('disciplineType and teachers');
```
Comment (basarabst) [inline src/v2/api/services/ScheduleService.ts:513]: split exception into 2 for readability

**Review by basarabst — CHANGES_REQUESTED** [review]

**Review by basarabst — APPROVED** [review]

**Review by SharpDevOps10 — APPROVED** [review]

**Review by just-hoshinon — APPROVED** [review]

**Pull request merged**